Persist runtime data in LittleFS

SW-Niko opened the pull request:
**Read and write runtime data persistent on LittleFS**

- The data is stored in JSON format.
- The data is written during WebApp 'OTA firmware upgrade' and during Webapp 'Reboot'
- For security reasons such as 'unexpected power cycles' or 'physical resets', data is also written once a day at 00:05
- The data will not be written if the last write operation was less than 10 Minutes ago.
- Thread save access to the data is provided by a mutex/lock.
- Indicate the save count and the last save time in the web UI

<img width="350" height="100" alt="grafik" src="https://github.com/user-attachments/assets/73e03155-1a88-44fd-b151-f4ed62c7f776" />


**I should also mention the following points:**

- The mutex only protects the RuntimeData class's own data
- I simply attached it to the OTA update callback and reset callback function to save the runtime data

See issue #1676

**To quickly test the runtime functionality on your system, follow these steps:**

- Update your system with the new firmware.
- Go to "Info" -> "System" and check "Runtime data save count / time       0 / no time"
- Force the system to save the runtime data on a JSON-File by doing a reboot via the web interface.
- Again go to "Info" -> "System" and check "Runtime data save count / time      1 / 15-Oct 19:33"
- You can also download the JSON-File and review its content.
```

{
    "info": {
        "version": 0,
        "save_count": 1,
        "save_epoch": 1760549624
    }
}
```

SW-Niko committed:
Runtime: store and load runtime data

- save and load runtime data persistent in a file
- read mode ON_DEMAND or ON_STARTUP
- the data is stored in JSON format
- the data is written during WebApp 'OTA firmware upgrade' and during Webapp 'Reboot'
- for security reasons such as 'unexpected power cycles' or 'physical resets', data is also written once a day at 00:05

SW-Niko committed:
Runtime: store and load runtime data

- save and load runtime data persistent in a file
- read mode ON_DEMAND or ON_STARTUP
- the data is stored in JSON format
- the data is written during WebApp 'OTA firmware upgrade' and during Webapp 'Reboot'
- for security reasons such as 'unexpected power cycles' or 'physical resets', data is also written once a day at 00:05

AndreasBoehm commented:
@SW-Niko Das geht leider nicht, da deine RuntimeData branch nicht im hoylabs repo sondern in deinem Fork liegt.

Du kannst `development` als base für deinen PR nehmen und erwähnen das er auf diesem hier basiert. Anders können wir das mit einem Fork leider nicht lösen.

SW-Niko commented:
Hello @schlimmchen or @AndreasBoehm ,
Do you see any chance of reviewing this PR in the near future?
I've expanded the Battery Guard with:
- Write the calculated DC-Pulse value into the runtime data
- Stop-Voltage Power Limiter
- Recharge Helper

But I need this PR and #2298 first to avoid triple nested PRs. 😀

SW-Niko committed:
Runtime: integration

- into update, reset
- WebUI

SW-Niko committed:
Runtime: integration

- into update, reset
- WebUI

SW-Niko committed:
Runtime: fix findings

- mutex no longer held during the entire file I/O
- _writeOK indicates successful write operation

SW-Niko committed:
Runtime: fix findings

- mutex no longer held during the entire file I/O
- _writeOK indicates successful write operation

AndreasBoehm commented on src/RuntimeData.cpp:
```diff
@@ -0,0 +1,264 @@
+    bool readOk = false;
+    JsonDocument doc;
+
+    // Note: We do not exit on read or allocation errors. In that case we need the default values
```
Can we use optionals instead of default values?
I could imagine that we can not set meaningful default values for a lot of upcoming cases

SW-Niko replied:
Not sure if I got you right?

The comment means ... If the runtime file does not exist, we do not change the values from the configuration or the default values.

For example: The internal resistance of the battery. If the runtime file does not exist or the internal resistance is not part of the runtime file, we use the default value from the configuration.

AndreasBoehm replied:
Lets stick to your example.

we currently have fileVersion `1` which does not contain the internal resistance of the battery.
What will happen when we read the file with fileVersion `1` but code version `2`? 
We set `_batteryInternalResistance` to a default value?

AndreasBoehm replied:
I am closing this conversation because it duplicates what i am saying in the other discussion in the same file.

AndreasBoehm commented on src/RuntimeData.cpp:
```diff
@@ -0,0 +1,264 @@
+    if (mode == ReadMode::START_UP) {
+        ; // todo: deserialize additional runtime data that can be initialized during startup
+    } else {
+        ; // todo: deserialize additional runtime data that can not be initialized during startup
```
What data to you have in mind that can not be read/initialized during startup?

SW-Niko replied:
Data that uses allocated memory. Like `auto upInv = PowerLimiterInverter::create(invConfig);`
So far I have only encountered one case... The correction list for inverter transmission in the case of the 24-VDC problem.

AndreasBoehm replied:
Maybe i am not getting how this will work in the end.

If i trigger `read()` what will happen?
`RuntimeClass` will access e.g. `PowerLimiterInverter` and set a value on it?

AndreasBoehm replied:
I found a example in a different PR of yours: https://github.com/hoylabs/OpenDTU-OnBattery/pull/2372/changes#diff-783e5c69ea25ae8ae20e5bd1f5faf7895f7ad15abebdb4ead79be611ed91ea23R174

I don't like the idea of injecting stored data into other objects.
Other objects should read from RuntimeData when they need the data.

Other objects will be tightly coupled to the RuntimeData class/object with the current approach.

AndreasBoehm replied:
I talke to claude about it and this is the result.

It violates both **Single Responsibility** (the class handles I/O *and* orchestrates restoration) and **Open/Closed** (not open for extension without modification).

**inversion of control via an interface/protocol**

Each subsystem should own its own serialization logic and register itself with a generic runtime-data manager. The manager only knows about the interface, not the concrete classes.

```cpp
// Interface each subsystem implements
class IRuntimeDataProvider {
public:
    virtual std::string getId() const = 0;
    virtual JsonObject serialize() const = 0;
    virtual void deserialize(const JsonObject& data) = 0;
    virtual ~IRuntimeDataProvider() = default;
};

// Generic manager — knows nothing about PowerLimiter etc.
class RuntimeDataManager {
public:
    void registerProvider(IRuntimeDataProvider* provider) {
        _providers[provider->getId()] = provider;
    }

    void saveAll(const std::string& path) {
        JsonDocument doc;
        for (auto& [id, provider] : _providers)
            doc[id] = provider->serialize();
        writeJsonToDisk(doc, path);
    }

    void loadAll(const std::string& path) {
        JsonDocument doc = readJsonFromDisk(path);
        for (auto& [id, provider] : _providers)
            if (doc.containsKey(id))
                provider->deserialize(doc[id]);
    }

private:
    std::map<std::string, IRuntimeDataProvider*> _providers;
};
```

Then each subsystem is self-contained:

```cpp
class PowerLimiter : public IRuntimeDataProvider {
public:
    std::string getId() const override { return "powerlimiter"; }
    JsonObject serialize() const override { /* pack own state */ }
    void deserialize(const JsonObject& data) override { /* restore own state */ }
};
```

And wiring it up happens at the composition root (e.g. `main` or your app init), not inside the manager:

```cpp
manager.registerProvider(&powerLimiter);
manager.registerProvider(&battery);
manager.registerProvider(&solar);
```

**What you gain:**
- `RuntimeDataManager` never changes when you add a subsystem
- Each subsystem owns its own data schema — no leakage
- Trivially testable in isolation
- Ordering, optional providers, missing keys on load all handled generically

This pattern fits well with embedded/OpenDTU-style architectures where subsystems are long-lived singletons — registration in the constructor or an `init()` call works cleanly.

SW-Niko replied:
I understand your concerns.

**RuntimeDataManager never changes when you add a subsystem**
I fully agree. (I just reused the pattern from the config)

**Each subsystem owns its own data schema — no leakage**
This is already the way of working. Hmmm ... maybe claude did not get it because this PR is only the shell?

**Trivially testable in isolation**
Ok

**Ordering, optional providers, missing keys on load all handled generically**
Ok, understood:

I will need some time to implement the new concept. 🤔

AndreasBoehm replied:
Let me think about this for some more time before you change something

AndreasBoehm replied:
Let's get this PR merged. We can discuss the interface situation with the first PR that makes use of the runtime data :)

AndreasBoehm commented on src/main.cpp:
```diff
@@ -145,12 +146,17 @@ void setup()
     Battery.init(scheduler);
+    // ... and here (before RuntimeData)
+
+    // Must be done after all other components have been initialized
```
why initialize it last? I thought that components could 'depend' on runtimedata at some point ant because of that we need to initialize it first?

SW-Niko replied:
The runtime file uses a very simple approach.

1. The class must be initialized with default or configuration values.
2. We update the values with values from the runtime file, if available.

It would be possible to do the other way but that would require more code and memory.

However... we need to be prepared before making decisions based on runtime data.

AndreasBoehm replied:
Makes sense

AndreasBoehm replied:
I need to take back my previous comment. Looking at the other dicussion within this PR it might be necessary to initialize RuntimeData first.
